pkg/hashreader/hashreader.go

```diff
@@ -0,0 +1,120 @@
+package hashreader
+
+// HashReader writes what it reads from an io.Reader to an MD5 and SHA256 hash.Hash.
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/hex"
+	"hash"
+	"io"
+)
+
+// HashReader verifies that the content of the io.Reader matches the expected checksums.
+type HashReader struct {
+	md5sum, sha256sum []byte // Byte values of md5sum, sha256sum of client sent values.
+
+	src                 io.Reader
+	size                int64
+	md5Hash, sha256Hash hash.Hash
+}
+
+// New returns a new HashReader computing the MD5 sum and SHA256 sum
+// (if set) of the provided io.Reader.
+func New(src io.Reader, size int64, md5Hex, sha256Hex string) (*HashReader, error) {
+	sha256sum, err := hex.DecodeString(sha256Hex)
+	if err != nil {
+		return nil, err
+	}
+	md5sum, err := hex.DecodeString(md5Hex)
+	if err != nil {
+		return nil, err
+	}
+	if hr, ok := src.(*HashReader); ok {
+		return hr, nil
+	}
+	var (
+		md5Hash    hash.Hash
+		sha256Hash hash.Hash
+	)
+	if len(md5sum) != 0 {
+		md5Hash = md5.New()
+	}
+	if len(sha256sum) != 0 {
+		sha256Hash = sha256.New()
+	}
+	return &HashReader{
+		md5sum:     md5sum,
+		sha256sum:  sha256sum,
+		src:        io.LimitReader(src, size),
+		size:       size,
+		md5Hash:    md5Hash,
+		sha256Hash: sha256Hash,
+	}, nil
+}
+
+func (r *HashReader) Read(p []byte) (n int, err error) {
+	n, err = r.src.Read(p)
```

Comment:
`n > 0` may be true in `err != nil`.  I think we need to fix this case as well.

Reply:
The case doesn't need to be handled if Read returns any other error other than io.EOF or nil then its a catastrophic.

Reply:
> An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF. 

This line explains this.

Reply:
Here is the text from https://golang.org/pkg/io/#Reader
```
When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF. 
```

Reply:
Correct but we don't need to do anything @balamurugana we already handle this.. it doesn't matter if n > 0 and err is anything other than io.EOF or nil it is a catastrophic error.  We only need to handle things for correct cases which are only two i.e either io.EOF or nil. 

So Reader implementations if n > 0 always `return either err == EOF or err == nil`  any other error is considered catastrophic and should be quickly failed.